File: app/src/main/java/com/klinker/android/twitter_l/data/roomdb/Hashtag.java
package com.klinker.android.twitter_l.data.roomdb;


import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.Index;
import androidx.room.PrimaryKey;

@Entity(tableName = "hashtags", indices = @Index(value= "name", unique = true))
public class Hashtag {

    @PrimaryKey(autoGenerate = true)
    public final Long id;

    @ColumnInfo(name="name")
    public String tag;

    //maybe add account?

    public Hashtag(long id, String tag) {
        this.id = id;
        this.tag = tag;
    }

    @Ignore
    public Hashtag(String tag) {
        this.id = null;
        this.tag = tag;
    }

}

File: app/src/main/java/com/klinker/android/twitter_l/data/roomdb/Hashtag.kt
package com.klinker.android.twitter_l.data.roomdb

import androidx.room.*

@Entity(tableName = "hashtags", indices = [Index(value = ["tag"], unique = true)])
data class Hashtag @JvmOverloads constructor(
        val tag: String,
        @PrimaryKey(autoGenerate = true) val id: Long = 0
) : BaseDao.TalonEntity<Hashtag> {

    override fun copyWithId(id: Long): Hashtag {
        return this.copy(id = id)
    }
}

@Dao
abstract class HashtagDao : BaseDao<Hashtag>() {

    @Query("SELECT * FROM hashtags WHERE tag LIKE :pattern")
    protected abstract fun queryPattern(pattern: String) : List<Hashtag>

    @Query("DELETE FROM hashtags WHERE id NOT IN (SELECT id FROM hashtags ORDER BY id DESC LIMIT :size)")
    abstract fun trimDatabase(size: Int)

    fun getHashtags(tag: String) : List<Hashtag> {
        return queryPattern("%$tag%")
    }
}